### PR TITLE
Add in metrics for detecting Redundant Pulls

### PR DESCRIPTION
**Pull request by gregcusack**
#### Problem
We had previously added in a metric for tracking gossip push messages through the network in [PR: #32725](https://github.com/solana-labs/solana/pull/32725). However, this metric does not account for redundant pull requests. 
*Redundant Pull:* A node receives a message via `PullResponse` and then receives the same message via `Push`. 
Redundant Pulls prevent us from accurately calculating how well messages are propagating via `Push`. 

#### Summary of Changes
Add in a metric to report when we receive a `Push` for a message we already (and first) received via `PullResponse`
- `num_push_dups` changed to `num_push_recv` and now tracks the number of times we have received a push message. 
- add a new metric to `cluster_info_crds_stats` called `num_redundant_pull_responses`. It counts the number of times a unique message is received via a Redundant Pull

#### Identifying redundant Pulls:
1) Receive a message via `PullRequest` that successfully updates `crds.table`, set the `num_push_recv` of this message to 0. Set `num_push_recv` to 1 if it is a `PushMessage`
2) Receive the same message again but via `Push`, it will fail to insert. Since the already existing entry has `num_push_recv  == 0`,  we know this is a Redundant Pull. 
3) Increment `CrdsStats.num_redundant_pull_responses`

#### Calculating fraction of messages received via Redundant Pull:
```
num_redundant_pull_responses / (num_redundant_pull_responses + all-push)
```

#### Monogon Simulation: % of redundant pulls in a 100 validator cluster
We see a mean Redundant Pull percentage of ~0.2%. But it does seem to increase with the number of validators
Screenshot 2024-03-11 at 8 09 13 PM

### CODE REVIEW

File: gossip/src/crds_gossip_push.rs [hunk 1]
```diff
@@ -157,6 +157,12 @@ impl CrdsGossipPush {
                         received_cache.record(origin, from, /*num_dups:*/ usize::MAX);
                         self.num_old.fetch_add(1, Ordering::Relaxed);
                     }
+                    Err(CrdsError::RedundantPull) => {
+                        // rewarded for being after Pull but first Push??
+                        received_cache.record(origin, from, /*num_dups:*/ 0);
+                        // still track that it is old message
+                        self.num_old.fetch_add(1, Ordering::Relaxed);
+                    }
```
Comment (gregcusack) [inline gossip/src/crds_gossip_push.rs]: Question: should the peer that sends the first push message that triggers the discovery of a redundant pull be treated the same in `ReceivedCache` as the peer who sends the first push message resulting in a successful update of the `crds.table`?  I would argue the answer is yes they should be treated the same:
Reasoning: a peer shouldn't be punished just because a `PullRequest` happened to beat it in reporting a datapoint. This also means the second peer would still gets its scored increased in `ReceivedCache` since `NUM_DUPS_THRESHOLD == 2`. Even though the peer may be slow it is still faster than any peer sending this same message after it. 

Possible argument against: Technically the peer wasn't the first one and this is a duplicate. There may be a reason besides chance that a `PullRequest` beat the `Push`.

File: gossip/src/crds.rs [hunk 2]
```diff
@@ -147,14 +149,20 @@ impl Cursor {
...
 
 impl VersionedCrdsValue {
-    fn new(value: CrdsValue, cursor: Cursor, local_timestamp: u64) -> Self {
+    fn new(value: CrdsValue, cursor: Cursor, local_timestamp: u64, route: GossipRoute) -> Self {
         let value_hash = hash(&serialize(&value).unwrap());
+        let num_push_recv = match route {
+            // If we received a PushMessage, set
+            // num_push_recv <- 1
+            GossipRoute::PushMessage(_) => 1u8,
+            _ => 0u8,
```
Comment (gregcusack) [inline gossip/src/crds.rs]: do we have to pay attention to `LocalMessage` or `PullRequest` here? At the very least I think we can ignore `PullRequest` since we're not updating our `crds.table` when we receive a `PullRequest`

Comment (behzadnouri) [reply gossip/src/crds.rs]: > we're not updating our crds.table when we receive a PullRequest

We actually upsert the origin of `PullRequest` into crds table:
https://github.com/anza-xyz/agave/blob/0f1ca20d3/gossip/src/crds_gossip_pull.rs#L296

My initial thought was that we can just leave it zero anyways, but looks like it might break `record_redundant_pull` below.

An ugly workaround would be define:
```rust
num_push_recv: Option<u8>,
```

Then in `VersionedCrdsValue::new`:
```rust
let num_push_recv = match route {
   GossipRoute::LocalMessage => None,
   GossipRoute::PullRequest => None,
   GossipRoute::PullResponse => Some(0),
   GossipRoute::PushMessage => Some(1),
}
```
Then `Some(0)` will accurately identify redundant pulls.

File: gossip/src/crds.rs [hunk 3]
```diff
@@ -303,8 +311,21 @@ impl Crds {
                     Err(CrdsError::InsertFailed)
                 } else if matches!(route, GossipRoute::PushMessage(_)) {
                     let entry = entry.get_mut();
-                    entry.num_push_dups = entry.num_push_dups.saturating_add(1);
-                    Err(CrdsError::DuplicatePush(entry.num_push_dups))
+                    // If num_push_recv == 0, we know this data was first received
+                    // via PullResponse. This is a redundant pull, meaning we received
+                    // a PushMessage after we had already received the same message
+                    // first via PullResponse
```
Comment (behzadnouri) [inline gossip/src/crds.rs]: I _think_ this is not true if the value was upserted from `GossipRoute::LocalMessage` or `GossipRoute::PullRequest` as in:
https://github.com/anza-xyz/agave/blob/0f1ca20d3/gossip/src/crds_gossip_pull.rs#L296
So this will over count redundant pull messages.

File: gossip/src/crds_gossip_push.rs [hunk 4]
```diff
@@ -157,6 +157,12 @@ impl CrdsGossipPush {
                         received_cache.record(origin, from, /*num_dups:*/ usize::MAX);
                         self.num_old.fetch_add(1, Ordering::Relaxed);
                     }
+                    Err(CrdsError::RedundantPull) => {
```
Comment (behzadnouri) [inline gossip/src/crds_gossip_push.rs]: If we are not distinguishing between `CrdsError::DuplicatePush(0)` and `CrdsError::RedundantPull` maybe we don't need to add `CrdsError::RedundantPull` anyways.
The `0` in `CrdsError::DuplicatePush(0)` should indicate this was received from push after it was upserted from pull-responses.

File: gossip/src/crds_gossip_push.rs (hunk 1 above)
Comment (behzadnouri) [reply gossip/src/crds_gossip_push.rs]: > should the peer that sends the first push message that triggers the discovery of a redundant pull be treated the same in ReceivedCache as the peer who sends the first push message resulting in a successful update of the crds.table?

yes, I think so.

> Possible argument against: Technically the peer wasn't the first one and this is a duplicate. There may be a reason besides chance that a PullRequest beat the Push.

I can't think of a good reason here. Basically if there was no outgoing pull-request this value would have been updated from push anyways.

File: gossip/src/crds.rs [hunk 5]
```diff
@@ -746,6 +767,10 @@ impl CrdsStats {
             GossipRoute::PullResponse => self.pull.record_fail(entry),
         }
     }
+
+    fn record_redundant_pull(&mut self) {
+        self.redundant_pull += 1;
+    }
```
Comment (behzadnouri) [inline gossip/src/crds.rs]: feels a bit unnecessary to add a fn for a one-liner which is only used one place.

File: gossip/src/crds_gossip_push.rs (hunk 1 above)
Comment (gregcusack) [reply gossip/src/crds_gossip_push.rs]: ya i agree

File: gossip/src/crds_gossip_push.rs (hunk 4 above)
Comment (gregcusack) [reply gossip/src/crds_gossip_push.rs]: oh you are right.

File: gossip/src/crds.rs (hunk 5 above)
Comment (gregcusack) [reply gossip/src/crds.rs]: ok i can just increment directly

File: gossip/src/crds.rs (hunk 2 above)
Comment (gregcusack) [reply gossip/src/crds.rs]: ahh shoot missed that. ya not an ideal option but would work. i'll think...

**Comment by gregcusack** [comment]
I ended up wrapping num_push_recv into a struct to make it a little more clear what is happening in the actual `crds.insert(..)` block. Similar sentiment as to what you had here but without an Option<u8>: https://github.com/anza-xyz/agave/pull/199#discussion_r1521537849
```
impl NumPushRecv {
    fn new(route: GossipRoute) -> Self {
        let (count, received_first_via_pull_request) = match route {
            GossipRoute::PullRequest => (0, true),
            GossipRoute::PushMessage(_) => (1, false),
            _ => (0, false),
        };
        Self {
            count,
            received_first_via_pull_response: received_first_via_pull_request,
        }
    }

    // If first time message was received via PullResponse and count == 0,
    // we know this data was first received via PullResponse.
    // This is a redundant pull, meaning we received a PushMessage
    // after we had already received the same message first via PullResponse
    fn is_redundant_pull(&self) -> bool {
        self.received_first_via_pull_response && self.count == 0
    }
}
```

File: gossip/src/crds.rs [hunk 6]
```diff
@@ -75,7 +75,7 @@ pub struct Crds {
     votes: BTreeMap<u64 /*insert order*/, usize /*index*/>,
     // Indices of EpochSlots keyed by insert order.
     epoch_slots: BTreeMap<u64 /*insert order*/, usize /*index*/>,
-    // Indices of DuplicateShred keyed by insert order.
+    // Indice of DuplicateShred keyed by insert order.
```
Comment (behzadnouri) [inline gossip/src/crds.rs]: was this change intentional?

File: gossip/src/cluster_info_metrics.rs [hunk 7]
```diff
@@ -639,6 +639,7 @@ pub(crate) fn submit_gossip_stats(
         ),
         ("RestartHeaviestFork-push", crds_stats.push.counts[13], i64),
         ("RestartHeaviestFork-pull", crds_stats.pull.counts[13], i64),
+        ("RedundantPull", crds_stats.redundant_pull, i64),
```
Comment (behzadnouri) [inline gossip/src/cluster_info_metrics.rs]: Can we actually move this next to other `pull_response_...` metrics?
https://github.com/anza-xyz/agave/blob/81075e60b/gossip/src/cluster_info_metrics.rs#L293-L317
We can name it something like `num_redundant_pull_responses`.

File: gossip/src/crds.rs [hunk 8]
```diff
@@ -115,6 +115,35 @@ pub(crate) struct CrdsDataStats {
 pub(crate) struct CrdsStats {
     pub(crate) pull: CrdsDataStats,
     pub(crate) push: CrdsDataStats,
+    pub(crate) redundant_pull: i64,
```
Comment (behzadnouri) [inline gossip/src/crds.rs]: `num_redundant_pull_responses` is probably more descriptive.
Also maybe add a comment what does this metric mean.

Also `u64` is a better type since this value is not meant to be negative.

File: gossip/src/crds.rs [hunk 9]
```diff
@@ -115,6 +115,35 @@ pub(crate) struct CrdsDataStats {
...
+    pub count: u8,
+    pub received_first_via_pull_response: bool,
+}
+
+impl NumPushRecv {
+    fn new(route: GossipRoute) -> Self {
+        let (count, received_first_via_pull_response) = match route {
+            GossipRoute::PullRequest => (0, true),
+            GossipRoute::PushMessage(_) => (1, false),
+            _ => (0, false),
```
Comment (behzadnouri) [inline gossip/src/crds.rs]: The problem with `_ => ...` is that if a new variant is added to the `GossipRoute` enum, then this code silently ignores the new variant which might not be the right thing.
Whereas if you list all the remaining variants explicitly, you get a compile error if a new variant is added to the enum, which would require to explicitly handle that new variant here.
https://rust-lang.github.io/rust-clippy/master/index.html#/wildcard_enum_match_arm

File: gossip/src/crds.rs [hunk 10]
```diff
@@ -115,6 +115,35 @@ pub(crate) struct CrdsDataStats {
...
+        }
+    }
+
+    // If first time message was received via PullResponse and count == 0,
+    // we know this data was first received via PullResponse.
+    // This is a redundant pull, meaning we received a PushMessage
+    // after we had already received the same message first via PullResponse
+    fn is_redundant_pull(&self) -> bool {
+        self.received_first_via_pull_response && self.count == 0
+    }
```
Comment (behzadnouri) [inline gossip/src/crds.rs]: I feel like this is a bit overkill/verbose for a metric.
Can we just use an `Option<u8>` to avoid verbosity for now and revise later if that was not sufficient?

File: gossip/src/crds.rs (hunk 6 above)
Comment (gregcusack) [reply gossip/src/crds.rs]: no it was not 🤦

File: gossip/src/crds.rs (hunk 9 above)
Comment (gregcusack) [reply gossip/src/crds.rs]: this is a very good point. thanks for the info. will avoid in the future (and fix now)

File: gossip/src/crds.rs (hunk 10 above)
Comment (gregcusack) [reply gossip/src/crds.rs]: ya that was my worry. no problem.

File: gossip/src/crds.rs [hunk 11]
```diff
@@ -303,8 +313,19 @@ impl Crds {
...
+                    let num_push_recv = match entry.num_push_recv {
+                        Some(count) => {
+                            if count == 0 {
+                                self.stats.lock().unwrap().num_redundant_pull_responses += 1;
+                            }
+                            Some(count.saturating_add(1))
+                        }
+                        None => Some(1),
+                    };
+                    let num_push_dups = num_push_recv.unwrap_or(0).saturating_sub(1);
```
Comment (gregcusack) [inline gossip/src/crds.rs]: I believe we could just use `unwrap()` here instead of `unwrap_or(0)` since `num_push_recv` will not be `None` after the `match` statement

File: gossip/src/crds.rs [hunk 12]
```diff
@@ -115,6 +115,9 @@ pub(crate) struct CrdsDataStats {
 pub(crate) struct CrdsStats {
     pub(crate) pull: CrdsDataStats,
     pub(crate) push: CrdsDataStats,
+    /// number of times a message was first received via a PullResponse
+    /// and that message was later received via a PushMessage
+    pub(crate) num_redundant_pull_responses: u64,
```
Comment (gregcusack) [inline gossip/src/crds.rs:120]: should I make this a `Counter` or leave it as `u64`?

Comment (behzadnouri) [reply gossip/src/crds.rs:120]: `u64` seems sufficient.

File: gossip/src/crds.rs (hunk 11 above)
Comment (behzadnouri) [reply gossip/src/crds.rs]: I think this could be written as:
```rust
if entry.num_push_recv == Some(0) {
   self.stats.lock().unwrap().num_redundant_pull_responses += 1;
}
let num_push_dups = entry.num_push_recv.unwrap_or_default();
entry.num_push_recv = Some(num_push_dups.saturating_add(1));
```

File: gossip/src/crds.rs [hunk 13]
```diff
@@ -127,8 +130,8 @@ pub struct VersionedCrdsValue {
     pub(crate) local_timestamp: u64,
     /// value hash
     pub(crate) value_hash: Hash,
-    /// Number of times duplicates of this value are recevied from gossip push.
-    num_push_dups: u8,
+    /// number of times this value is recevied via PushMessage
+    num_push_recv: Option<u8>,
```
Comment (behzadnouri) [inline gossip/src/crds.rs:136]: Maybe add a comment here that
* `None` indicates the value was upserted by `GossipRoute::{LocalMessage,PullRequest}`.
* `Some(0)` indicates the value was upserted by `GossipRoute::PullResponse`.
* `Some(k) if k > 0` indicates the value was upserted by `GossipRoute::PushMessage` with `k - 1` push duplicates.

**Review by behzadnouri — APPROVED** [review]
lgtm

**Comment by behzadnouri** [comment]
I think it would be good to backport this to v1.18 to both get the metrics, and also fix the num-dups accounting for received-cache scoring.

**Comment by gregcusack** [comment]
> also fix the num-dups accounting for received-cache scoring.

what do you mean by this? isn't the scoring the same from the ReceivedCache side of things?

**Comment by behzadnouri** [comment]
> > also fix the num-dups accounting for received-cache scoring.
> 
> what do you mean by this? isn't the scoring the same from the ReceivedCache side of things?

Basically this: https://github.com/anza-xyz/agave/pull/199#discussion_r1520798636

**Comment by gregcusack** [comment]
> > > also fix the num-dups accounting for received-cache scoring.
> > 
> > 
> > what do you mean by this? isn't the scoring the same from the ReceivedCache side of things?
> 
> Basically this: [#199 (comment)](https://github.com/anza-xyz/agave/pull/199#discussion_r1520798636)

ok I see. ya makes sense. thank you!
